src/main/java/com/google/sps/data/Journey.java

```diff
@@ -56,12 +59,235 @@ public Pair(int first, int second) {
+      orderedWaypoints.add(waypoints.get(currentWaypoint));
+      int prevState = currentState ^ (1 << currentWaypoint);
+      int prevWaypoint = bestTime.get(currentState).get(currentWaypoint).second;
+      currentWaypoint = prevWaypoint;
+      currentState = prevState;
+    }
+    // add the start of the journey at the end of orderedWaypoints.
+    orderedWaypoints.add(start);
+    Collections.reverse(orderedWaypoints);
+    return orderedWaypoints;
+  }
+
+  /**
+   * Creates and initializes bestTime matrix, a matrix where bestTime[config][lastWaypoint] = the
+   * minimum time to visit all points with bits set to 1 in config such that the last visited point
+   * is lastWaypoint; config is a bitmask where bits set to 1 represent currently visited waypoints.
+   */
+  private ArrayList<ArrayList<Pair>> initializeBestTimeMatrix() throws ApiException, IOException, InterruptedException, DataNotFoundException {
+    ArrayList<ArrayList<Pair>> bestTime = new ArrayList<>();
+    int numberOfWaypoints = getNumberOfWaypoints();
+    for (int config = 0; config < (1 << numberOfWaypoints); ++config) {
+      // add the config-th line to the matrix with numberOfWaypoints columns of Integer.MAX_VALUE
+      ArrayList<Pair> currentRow = new ArrayList<Pair>();
+
+      for (int waypointIndex = 0; waypointIndex < numberOfWaypoints; ++ waypointIndex) {
+        currentRow.add(new Pair(Integer.MAX_VALUE, -1));
+      }
+      bestTime.add(currentRow);
+    }
+
+    int waypointIndex = 0;
+    for (CourierStop waypoint : waypoints) {
+      Pair currentElement = bestTime.get(1 << waypointIndex).get(waypointIndex);
+      currentElement.first = pathFinder.getTimeInSecondsBetweenPoints(start.getPoint(), waypoint.getPoint());
+      waypointIndex += 1;
+    }
+
+    return bestTime;
+  }
+
+  /**
+   * Computes and returns the bestTime matrix such that bestTime[config][latsWaypoint].first is the
+   * min time needed to visit all points with bits set to 1 in config such that the path ends in 
+   * lastWaypoint and bestTime[config][latsWaypoint].second represents the point that was visited 
+   * right before lastWaypoint. 
+   *
+   * The method will call initializeBestTimeMatrix() to set the initial states for bestTime.
+   * for each configuration config from 1 to (1<<(numberOfWaypoints - 1)) - 1, AND 
+   *   for each waypointIndex i from 0 to numberOfWaypoints - 1 that is not set to 1 in config
+   *   AND there is no index k set to 1 in config such that waypoint k must be visited after i
+   *     newConfig = config + (1<<i), i.e. newConfig has bits in config plus the i-th bit set to 1
+   *     update bestTime[newConfig][i] with each matrix element bestTime[config][j] for which
+   *     j is set to 1 in config, i.e. consider each index j set to 1 in config as the last visited
+   *     waypoint before i and update bestTime[newConfig][i].first with bestTime[config][j].first +
+   *     the time to get from waypoint j to waypoint i.
```

Comment:
This might be just me, but phrase completely went over my head. I think it's just saying the same thing as the expressions below (which were much easier for me to digest for some reason) + the fact that it respects point restrictions?